File: Data Structures & Algorithms/binary-search/submission-2.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        i = 0
        j = len(nums) - 1

        while i <= j:
            mean = (i + j) // 2

            if target == nums[mean]:
                return mean
            elif target > nums[mean]:
                i = mean + 1
            else:
                j = mean - 1

        return -1

File: Data Structures & Algorithms/subsets/submission-2.ts
class TreeNode<TValue> {
    public left: TreeNode<TValue> | null
    public right: TreeNode<TValue> | null
    
    constructor(public val: TValue | null) {}
}

class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    subsets(nums: number[]): number[][] {
        const res: number[][] = []
        const subset: number[] = []

        const dfs = (i: number) => {
            if (i >= nums.length) {
                res.push(Array.from(subset))
                return
            }

            subset.push(nums[i])
            dfs(i + 1)

            subset.pop()
            dfs(i +1)
        }

        dfs(0)

        return res
    }
}
